### ISSUE TITLE
Long-polling for immediate updates?

### ISSUE THREAD

**Issue by jwater7** [pavoni/pyvera#82]
I was reading [this thread](http://forum.micasaverde.com/index.php?topic=27866.0), and it seems that vera has support for long-polling that would remove any delay in notifications or triggers.  I believe pyvera has been implemented partially with this, but it does include an additional sleep(1)
https://github.com/pavoni/pyvera/blob/1fd6aca819381e67ca8fdc54a732fe4ddd92c590/pyvera/subscribe.py#L132
which wouldn't be implementing a proper long-poll.  I believe either removing this sleep or only doing this sleep on http request errors would be beneficial.  Would like to hear your input/ideas on this.

[home-assistant](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/vera.py) is using pyvera-subscribe for device status as far as I can tell.

**Comment by pavoni** [comment]
So pyvera certainly uses the long poll code, and home assistant uses this.

Quite a while since I wrote it - but I spent quite a while trying to tune the timing parameters to get quick responses, but not to overload the very hub.

Vera events seem to cascade ( turning on a light produces a lot of power update events for example), so with a very short timeout you can get a lot of callbacks, and a lot of short polls seemed to cause problems for the hub.

If I remember I chose a fairly short timeout for the long poll (for when there are events) to get fast response, but added the sleep so that cascading events don’t cause too many polls.

**Comment by jwater7** [comment]
That makes sense, so if you chose to remove the sleep, then Vera has problems and set a longer min polling time, then you also wouldn’t get very quick updates.  It sounds like it’s as good as it can get with the current Vera api limitations but is kind of unfortunate that events in between that second won’t be able to trigger anything immediately.

Maybe a bit off topic, but do you have a system using Vera and Home Assistant?  If so, have you experienced any delay turning a light on from a motion sensor using home assistant automations for example?

**Comment by pavoni** [comment]
I’m not sure I set the numbers very scientifically- so do feel free to experiment.

I don’t use a motion to trigger anything - but I have a zwave switch that is only used to trigger a home assistant scene. That switches other devices in less than a second. Although switching it again can take 4/5 secs.  Once devices start being activated there are a lot of Vera events (you can turn Vera tracing on if you want to see the details).

**Comment by jwater7** [comment]
With the latest release (which includes commit 2ea1464c4909b5858772abe3950db44a2d1cdd5b), it looks like this issue is actually resolved (long-polling with sleep only on exception has been implemented) now in case anyone is interested.

**Comment by jwater7** [comment]
@toggledbits, there used to be 2 sleeps, one was on error (which makes a lot of sense I think) and the other one was a hard "sleep(1)" that you removed and that change is now in master.  Is that the one you're referring to as a safety net and are considering adding back in?  I'm not sure of the implications, but I'm more just curious if there's something identified as the culprit.

**Comment by toggledbits** [comment]
There are still two sleeps in the current code. One happens between polls when there is no error (whether or not there is delta data)--that's the 1 second sleep in the current code. It got moved, but it's still there. I had removed it in the initial commit of my pull, but @pavoni told me why he had it there, and testing confirmed it's better with it (if you want detailed info, let me know--it may be a little OT and long for this thread). There is also a 10-second sleep on any trapped exception. The 1-second sleep seems to be good throttling, and now that I've got my other little fix incorporated in my version (clear timestamp to None on exception--a one-line change before the 10-second delay), it runs like a clock.
